File: 프로그래머스/3/64062. 징검다리 건너기/징검다리 건너기.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int solution(vector<int> stones, int k) {
    int answer = 0;
    
    int max = *max_element(stones.begin(), stones.end());
    int min = 1;
    
    while(min <= max)
    {
        int mid = (min + max) / 2;
        
        int count = 1;
        for(int i = 0; i < stones.size(); ++i)
        {
            if (stones[i] - mid < 0)
            {
                ++count;
            }
            else
            {
                count = 1;
            }
            
            if (count > k)
            {
                break;
            }
        }
        
        if (count > k)
        {
            max = mid -1;
        }
        else
        {
            answer = mid;
            min = mid + 1;
        }
    }
    
    
    return answer;
}

File: 프로그래머스/3/64062. 징검다리 건너기/징검다리 건너기.java
import java.util.*;

class Solution {
    public int solution(int[] stones, int k) {
        int answer = 0;
        
        int start = 1;
        int end = 200_000_000;
        
        while(start <= end) {
            int mid = (start + end) / 2;
            int length = 1;
            int max = Integer.MIN_VALUE;
            
            for(int i = 0; i < stones.length; i++) {
                if(stones[i] - mid < 0) {
                    length++;
                } else {
                    length = 1;
                }
                max = Math.max(max, length);
            }
            
            if(max > k) {
                end = mid - 1;  
            } else {
                start = mid + 1;
                answer = mid;
            }
        }
        return answer;
    }
}